Kernel/FileSystem/FATFileSystem.cpp

```diff
@@ -0,0 +1,357 @@
+{
+    VERIFY(has_flag(m_entry.attributes, FATAttributes::Directory));
+
+    Vector<FATLongFileNameEntry> lfn_entries;
+    auto blocks = TRY(read_block_list());
+
+    for (u32 i = 0; i < blocks->size() / sizeof(FATEntry); i++) {
+        auto* entry = reinterpret_cast<FATEntry*>(blocks->data() + i * sizeof(FATEntry));
+        if (entry->filename[0] == end_entry_byte) {
+            dbgln_if(FAT_DEBUG, "FATFS: Found end entry");
+            return nullptr;
+        } else if (static_cast<u8>(entry->filename[0]) == unused_entry_byte) {
+            dbgln_if(FAT_DEBUG, "FATFS: Found unused entry");
+            lfn_entries.clear();
+        } else if (entry->attributes == FATAttributes::LongFileName) {
+            dbgln_if(FAT_DEBUG, "FATFS: Found LFN entry");
+            TRY(lfn_entries.try_append(*reinterpret_cast<FATLongFileNameEntry*>(entry)));
+        } else {
+            dbgln_if(FAT_DEBUG, "FATFS: Found 8.3 entry");
+            lfn_entries.reverse();
+            auto inode = TRY(FATInode::create(fs(), *entry, lfn_entries));
+            if (TRY(callback(inode)))
+                return inode;
+            lfn_entries.clear();
+        }
+    }
+
+    return EINVAL;
+}
+
+ErrorOr<NonnullOwnPtr<KString>> FATInode::compute_filename(FATEntry& entry, Vector<FATLongFileNameEntry> const& lfn_entries)
+{
+    if (lfn_entries.is_empty()) {
+        StringBuilder filename;
+        filename.append(byte_terminated_string(StringView(entry.filename, normal_filename_length), ' '));
+        if (entry.extension[0] != ' ') {
+            filename.append('.');
+            filename.append(byte_terminated_string(StringView(entry.extension, normal_extension_length), ' '));
+        }
+        return TRY(KString::try_create(filename.string_view()));
+    } else {
+        StringBuilder filename;
+        for (auto& lfn_entry : lfn_entries) {
+            filename.append(lfn_entry.characters1[0]);
+            filename.append(lfn_entry.characters1[1]);
+            filename.append(lfn_entry.characters1[2]);
+            filename.append(lfn_entry.characters1[3]);
+            filename.append(lfn_entry.characters1[4]);
+            filename.append(lfn_entry.characters2[0]);
+            filename.append(lfn_entry.characters2[1]);
+            filename.append(lfn_entry.characters2[2]);
+            filename.append(lfn_entry.characters2[3]);
+            filename.append(lfn_entry.characters2[4]);
+            filename.append(lfn_entry.characters2[5]);
+            filename.append(lfn_entry.characters3[0]);
+            filename.append(lfn_entry.characters3[1]);
+        }
+
+        return TRY(KString::try_create(byte_terminated_string(filename.string_view(), lfn_entry_text_termination)));
+    }
+
+    VERIFY_NOT_REACHED();
+}
+
+time_t FATInode::fat_date_time(u16 date, u16 time) const
+{
+    // Date format:
+    // - Year  7 bits
+    // - Month 4 bits
+    // - Day   5 bits
```

Comment:
This information might be better stored in a proper bit-field. That makes the code easier to read, and it saves us from potential mistakes while shifting/masking. Same for the "Time format" below.

Reply:
Is there an pre-existing structure / class that I could use for that?

Reply:
It would be something similar to [this](https://github.com/SerenityOS/serenity/blob/master/AK/FPControl.h#L41-L62), but obviously purpose-fit to the problem at hand.

I'd suggest the following (and a respective second definition for `FATPackedTime`):

```
union FATPackedDate {
    u16 value;
    struct {
        u16 year : 7;
        u16 month : 4;
        u16 day : 5; 
    };
};
static_assert(sizeof(FATPackedDate) == 16);
```

That way, you can access `packed_value.value` to read/write the raw value that the filesystem actually stores, and `packed_value.year`, `.month` and `.day` to read/write the parsed values.

This is also compatible with members inside structs, so you could define the various date and time members as that type directly (i.e. `FATPackedDate last_accessed_date` inside of `FATEntry`);